### PR TITLE
Add opentelemetry to mapper-framework for enhancing pushmethod.

### LINKED ISSUE #5478: Device data supports storage to prometheus
**What would you like to be added/modified**:
Our current business scenario is that there is a Prometheus service that supports rest api（http://xxxxxx/api/v1/write） writing, and we hope to store device data in it.

**Why is this needed**:

### PR DESCRIPTION
**Pull request by forsaken628**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add opentelemetry to mapper-framework for enhancing pushmethod db.

**Which issue(s) this PR fixes**:

Fixes #5478

**Special notes for your reviewer**:
similar as #5376

**Does this PR introduce a user-facing change?**:

```release-note
Add opentelemetry to mapper-framework for enhancing pushmethod db. 
```

### CODE REVIEW

**Comment by Shelley-BaoYue** [comment]
cc @wbc6080  
Could you please add more descriptive in the title of each commit?

**Review by wbc6080 — COMMENTED** [review]
Thank you very much for your contribution! But I have a few questions about this solution:
1.opentelemetry does not seem to be a database. I don’t quite understand why this part of the function should be placed in the database processing part of the mapper-framework data plane.
2. In addition to providing the function of pushing to the database, the current mapper-framework can also push data to user applications through http/mqtt. Why can't we directly access the prometheus api through http to complete data push, but need to use opentelemetry？

**Comment by forsaken628** [comment]
> 1.opentelemetry does not seem to be a database. I don’t quite understand why this part of the function should be placed in the database processing part of the mapper-framework data plane. 

fixed

>2. In addition to providing the function of pushing to the database, the current mapper-framework can also push data to user applications through http/mqtt. Why can't we directly access the prometheus api through http to complete data push, but need to use opentelemetry？

The current http implementation does not match this [remote_write_spec](https://prometheus.io/docs/concepts/remote_write_spec
)
```
The following headers MUST be sent with the HTTP request:

    Content-Encoding: snappy
    Content-Type: application/x-protobuf
    User-Agent: <name & version of the sender>
    X-Prometheus-Remote-Write-Version: 0.1.0

```

And why we should providing opentelemetry push?

https://opentelemetry.io/ecosystem/vendors/
https://opentelemetry.io/docs/collector/
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter

**Comment by Shelley-BaoYue** [comment]
Hi, @forsaken628   Would you be willing to attend KubeEdge community meeting (Wednesdays at 16:00-17:30 every week)? we can disscuss in detail about this feature.

**Comment by luomengY** [comment]
> opentelemetry

Is it possible to change the differences you mentioned based on the existing HTTP push methods? Here, the push method is for some databases, and it may not be necessary to add a separate opentelemetry push method.

**Comment by forsaken628** [comment]
I also don't think it's reasonable to add a specific push mechanism solely for Prometheus. However, the OpenTelemetry protocol is not only for support Prometheus but also for support various other database, including Apache SkyWalking, ClickHouse, GreptimeDB, Elastic, and so on. https://opentelemetry.io/ecosystem/vendors/

**Comment by wbc6080** [comment]
This is indeed a relatively large feature implementation. I think it is more appropriate to introduce it at a regular community meeting and see the opinions of other developers. @forsaken628  Do you have time to attend the regular meeting this Wednesday, May 29, from 4:00 to 5:30 pm? And could you show some materials such as proposal or other documentation to explain this part to everyone?

**Comment by forsaken628** [comment]
ok @wbc6080

**Comment by Shelley-BaoYue** [comment]
Hi, @forsaken628  I have added this topic in community meeting agenda https://docs.google.com/document/d/1Sr5QS_Z04uPfRbA7PrXr3aPwCRpx7EtsyHq7mp6CnHs/edit#heading=h.tzlhovcfulbj.  Welcome you to take part in meeting at 16:00 today, and the meeting link is https://zoom.us/j/4167237304

**Comment by JiaweiGithub** [comment]
Good job ! opentelemetry is a very popular monitoring framework currently. I think the method based on opentelemetry is very good. Others can also participate in this code to realize the integration of opentelemetry and other components.

**Review by Shelley-BaoYue — COMMENTED** [review]
Please run `hack/generate-crds.sh` and push the modified.

**Review by fisherxu — COMMENTED** [review]
@forsaken628 The lint error has been fixed in #5730, you can remove the last commit

File: pkg/apis/dmi/v1beta1/api.proto [hunk 1]
```diff
@@ -225,6 +225,7 @@ message CustomizedValue {
 message PushMethod {
     PushMethodHTTP http = 1;
     PushMethodMQTT mqtt = 2;
+    PushMethodOTEL otel = 4;
     DBMethod dbMethod = 3;
```
Comment (wbc6080) [inline pkg/apis/dmi/v1beta1/api.proto:229]:
```suggestion
    PushMethodOTEL otel = 3;
    DBMethod dbMethod = 4;
```
Would it be better if modified like this?

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+package otel
+
+import (
+	"context"
+	"time"
+
+	"github.com/kubeedge/Template/driver"
+	"github.com/kubeedge/mapper-framework/pkg/common"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"k8s.io/klog/v2"
```
Comment (wbc6080) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go:26]:
```suggestion
        "go.opentelemetry.io/otel/metric/instrument"
	"k8s.io/klog/v2"
	
	"github.com/kubeedge/Template/driver"
	"github.com/kubeedge/mapper-framework/pkg/common"
	
```
Need to adjust the order of import

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler_nostream.go [hunk 3]
```diff
@@ -1,3 +1,5 @@
+//go:build nostream
+
```
Comment (wbc6080) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/handler_nostream.go:2]: There is no need to add cross-compilation conditions, this part will be processed in the mapper-framework generation script

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/go.mod [hunk 4]
```diff
@@ -6,41 +6,59 @@ require (
 	github.com/eclipse/paho.mqtt.golang v1.2.0
 	github.com/go-redis/redis/v8 v8.11.5
 	github.com/go-sql-driver/mysql v1.7.1
-	github.com/golang/protobuf v1.5.3 // indirect
-	github.com/gorilla/mux v1.8.0 // indirect
 	github.com/influxdata/influxdb-client-go/v2 v2.13.0
 	github.com/kubeedge/kubeedge v0.0.0
-	github.com/kubeedge/mapper-framework v0.0.0
+	github.com/kubeedge/mapper-framework v0.0.0-00010101000000-000000000000
```
Comment (wbc6080) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/go.mod:11]: Why do you need to modify this dependency?

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/device/device.go [hunk 5]
```diff
@@ -149,14 +150,18 @@ func dataHandler(ctx context.Context, dev *driver.CustomizedDev) {
...
 		// handle push method
 		if twin.Property.PushMethod.MethodConfig != nil && twin.Property.PushMethod.MethodName != "" {
-			dataModel := common.NewDataModel(dev.Instance.Name, twin.Property.PropertyName, dev.Instance.Namespace, common.WithType(twin.ObservedDesired.Metadata.Type))
-			pushHandler(ctx, &twin, dev.CustomizedClient, &visitorConfig, dataModel)
+			switch twin.Property.PushMethod.MethodName {
+			case common.PushMethodOTEL:
+				otelMethod.DataHandler(ctx, &twin, dev.CustomizedClient, &visitorConfig, dataModel)
+			default:
+				pushHandler(ctx, &twin, dev.CustomizedClient, &visitorConfig, dataModel)
+			}
```
Comment (wbc6080) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/device/device.go:161]: Can `PushMethodOTEL` also be put into `pushHandler` for processing?

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/client.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+package otel
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/kubeedge/mapper-framework/pkg/common"
```
Comment (Shelley-BaoYue) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/client.go:26]: Please adjust the import sort.

File: pkg/apis/dmi/v1beta1/api.proto [hunk 7]
```diff
@@ -228,7 +228,8 @@ message CustomizedValue {
 message PushMethod {
     PushMethodHTTP http = 1;
     PushMethodMQTT mqtt = 2;
-    DBMethod dbMethod = 3;
+    PushMethodOTEL otel = 3;
+    DBMethod dbMethod = 4;
```
Comment (fisherxu) [inline pkg/apis/dmi/v1beta1/api.proto:232]: Changed the message number, there will be compatibility issue with older version.
ptal @WillardHu @Shelley-BaoYue @wbc6080 @JiaweiGithub 
/hold

**Review by Shelley-BaoYue — COMMENTED** [review]
Please rebase to fix conflict : )

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/metric/instrument"
+	"k8s.io/klog/v2"
+
+	"github.com/kubeedge/Template/driver"
+	"github.com/kubeedge/mapper-framework/pkg/common"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"k8s.io/klog/v2"
```
Comment (wbc6080) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go:29]: Duplicate reference to package
```suggestion
	"go.opentelemetry.io/otel/metric/instrument"
	"k8s.io/klog/v2"

	"github.com/kubeedge/Template/driver"
	"github.com/kubeedge/mapper-framework/pkg/common"
```

**Review by wbc6080 — COMMENTED** [review]
Thanks! cc @WillardHu @fisherxu 
/lgtm

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	meter := provider.Meter(meterName)
+
+	// gauge, err := meter.Float64ObservableGauge() // otel metric api 1.0+
+	gauge, err := meter.AsyncFloat64().Gauge(dataModel.PropertyName) // otel metric api 0.32
+	if err != nil {
+		klog.Errorf("create metric fail: %v", err)
+		return
+	}
+
+	_, err = meter.RegisterCallback(func(_ context.Context, o api.Observer) error {
```
Comment (WillardHu) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go:62]: err is not check here.

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+		data, err := client.GetDeviceData(visitorConfig)
+		if err != nil {
+			klog.Errorf("get device data fail: %v", err)
+			return err
```
Comment (WillardHu) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go:66]:
```golang
return fmt.Errorf("get device data fail: %v", err")
```
Return an error statck is better then print log.

File: staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+package otel
+
+import (
+	"context"
+	"time"
+
+	"github.com/kubeedge/Template/driver"
+	"github.com/kubeedge/mapper-framework/pkg/common"
+	"k8s.io/klog/v2"
```
Comment (WillardHu) [inline staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/publish/otel/handler.go:25]: Grouping imported packages
```golang
"context"
"time"

"k8s.io/klog/v2"

"github.com/kubeedge/Template/driver"
"github.com/kubeedge/mapper-framework/pkg/common"
```

**Pull request merged**